### ISSUE TITLE
Exercise 2.11: Vectorised function - Possible solution failing PyBryt checks

### ISSUE THREAD

**Issue by acse-5d697a4f** [ese-msc/introduction-to-python#78]
I have the following possible solution for the 2.11 exercise depicted [here](https://github.com/ese-msc/introduction-to-python/blob/256d6765c213ec375c75cd75e7c17f682a9a2971/lecture2/lecture2.ipynb#L3210):

```python
import pybryt
import math
import numbers
import numpy as np
from lecture import pybryt_reference

A33 = np.array([[0, 12, -1], [-1, -1, -1], [11, 5, 5]])

def f_cubic(A):
    res = np.zeros(A.shape)
    for idx, x in np.ndenumerate(A):
        res[idx] = x**3 + x * math.e**(x) + 1.
    return res

# Assertion Checks
assert isinstance(A33, np.ndarray)
assert all([isinstance(i, numbers.Real) for i in A33.flatten()])
assert A33.shape == f_cubic(A33).shape == (3, 3)
assert np.isclose(np.sum(A33), 29)
assert callable(f_cubic)
assert np.allclose(f_cubic(A33), np.power(A33, 3) + np.multiply(A33, np.exp(A33)) + 1)
assert all([isinstance(i, numbers.Real) for i in f_cubic(A33).flatten()])
assert np.allclose(f_cubic(np.array([0])), 1)

# PyBryt Checks
with pybryt.check(pybryt_reference(2, 11)):
    f_cubic(np.array([[1, 2, -6], [2, 2, -5]])), A33
```

Assertions pass, but PyBryt errors out with the following message:

```text
REFERENCE: exercise-2_11
SATISFIED: False
MESSAGES:
  - ERROR: You are not populating resulting vector correctly.
  - SUCCESS: You created a zero-matrix before looping.
  - ERROR: Hmmm... The result of your function is wrong.
  - SUCCESS: Your matrix A33 is correct.
```

While I recognize that the same issue was reported in #72, I'd still be interested in knowing what causes two correct solutions to yield errors with PyBryt. Would a floating-point precision/tolerance be the issue here? Since we don't have access to the reference solutions, it's not possible to explore the issue. So I would appreciate it if you could depict a possible reason for this issue. 

Note: I experienced a similar problem with another exercise (do not remember which anymore). Assertions were passing, but PyBryt was not. After inspecting, I noticed that my code was wrong and assertions were not covering all the scenarios.

**Comment by acse-ci223** [comment]
Adding to that, I'd like to add a more pythonic way in which I personally completed the exercise so that it's also added to PyBryt.


```python
A33 = np.array([[0,12,-1],[-1,-1,-1],[11,5,5]])

def f_cubic(A):
    def f(x):
        return x**3+x*math.e**x+1
    return np.vectorize(f)(A)

print(A33**3 + A33*math.e**A33+1)
print(f_cubic(A33))
```

PyBryt output is:

```
REFERENCE: exercise-2_11
SATISFIED: False
MESSAGES:
  - ERROR: You are not populating resulting vector correctly.
  - ERROR: You have not created a zero-matrix before looping..
  - ERROR: Hmmm... The result of your function is wrong.
  - SUCCESS: Your matrix A33 is correct.
```

**Comment by acse-js3623** [comment]
My simple code and output here, but can't pass the PyBryt as well. 

```
import numpy as np

A33 = np.array([[0, 12, -1],[-1, -1, -1],[11, 5, 5]])

def f_cubic(A):
  return A**3 + A*np.exp(A) + 1

f_cubic(A33)
```
Outputs:
```
array([[ 1.00000000e+00,  1.95478650e+06, -3.67879441e-01],
       [-3.67879441e-01, -3.67879441e-01, -3.67879441e-01],
       [ 6.59947559e+05,  8.68065796e+02,  8.68065796e+02]])

```

**Comment by marijanbeg** [comment]
Hi @acse-js3623 and @acse-5d697a4f, thank you very much for raising this issue. We have investigated this and the reason you were getting an error was a mistake in our reference solution.

To create a zero-matrix before populating it, we used `np.zeros_like(A)`, where `A` is the input matrix, instead of `np.zeros(A.shape)`. Since the input matrix is a matrix of integers, the created zero matrix will have the same `dtype` as matrix `A` - integer. Therefore, even though the formula results in a float, the fractional part was chopped and only an integer was placed into the results matrix.

We have modified our reference solution and both of your proposed solutions now pass PyBryt tests.
